File: golang/5-longest-palindromic-substring.go
package golang

func longestPalindrome(s string) string {
    startIndex, maxLength, n := 0, 0, len(s)

    checkPalindromes := func(left int, right int) {
        for left>=0 && right<n && s[left]==s[right] {
            currentLength := right-left+1
            if currentLength>maxLength {
                startIndex = left
                maxLength = currentLength
            }
            left--
            right++
        }
    }
    
    for i := 0; i < n; i++ {
        checkPalindromes(i, i)
        checkPalindromes(i, i+1)
    }
    
    return s[startIndex:startIndex+maxLength]
}

File: typescript/5-longest-palindromic-substring.ts
function longestPalindrome(s: string): string {
    let maxLen = 0;
    let start = 0;
    const checkPalindromes = (left: number, right: number) => {
        while (left >= 0 && right < s.length && s[left] === s[right]) {
            let len = right - left + 1
            if (len > maxLen) {
                start = left
                maxLen = len
            }
            left--
            right++
        }
    }
    
    for (let i = 0; i < s.length; i++) {
        checkPalindromes(i, i)
        checkPalindromes(i, i+1)  
    }
    
    return s.substr(start, maxLen)
};
